File: GXpert/GXpert.Web/Modules/Activation/CouponCode/CouponCodeForm.cs
using Serenity.ComponentModel;
using System;

namespace GXpert.Activation.Forms;

[FormScript("Activation.CouponCode")]
[BasedOnRow(typeof(CouponCodeRow), CheckNames = true)]
public class CouponCodeForm
{
    [HalfWidth]
    public string Code { get; set; }
    [HalfWidth]
    public int PlayListId { get; set; }
    [HalfWidth]
    public short ValidityType { get; set; }
    [HalfWidth]
    public short CountType { get; set; }
    [HalfWidth]
    public short Count { get; set; }
    [HalfWidth]
    public short ValidityInDays { get; set; }
    [HalfWidth]
    public DateTime ValidDate { get; set; }
    [HalfWidth]
    public int ConsumedCount { get; set; }
    [HalfWidth]
    public DateTime CouponValidityDate { get; set; }
}

File: GXpert/GXpert.Web/Modules/ServerTypes/Activation/CouponCodeForm.ts
import { StringEditor, LookupEditor, EnumEditor, IntegerEditor, DateEditor, PrefixedContext, initFormType } from "@serenity-is/corelib";
import { ECountType } from "../Web/Enums.ECountType";
import { EValidityType } from "../Web/Enums.EValidityType";

export interface CouponCodeForm {
    Code: StringEditor;
    PlayListId: LookupEditor;
    ValidityType: EnumEditor;
    CountType: EnumEditor;
    Count: IntegerEditor;
    ValidityInDays: IntegerEditor;
    ValidDate: DateEditor;
    ConsumedCount: IntegerEditor;
    CouponValidityDate: DateEditor;
}

export class CouponCodeForm extends PrefixedContext {
    static readonly formKey = 'Activation.CouponCode';
    private static init: boolean;

    constructor(prefix: string) {
        super(prefix);

        if (!CouponCodeForm.init)  {
            CouponCodeForm.init = true;

            var w0 = StringEditor;
            var w1 = LookupEditor;
            var w2 = EnumEditor;
            var w3 = IntegerEditor;
            var w4 = DateEditor;

            initFormType(CouponCodeForm, [
                'Code', w0,
                'PlayListId', w1,
                'ValidityType', w2,
                'CountType', w2,
                'Count', w3,
                'ValidityInDays', w3,
                'ValidDate', w4,
                'ConsumedCount', w3,
                'CouponValidityDate', w4
            ]);
        }
    }
}

[EValidityType, ECountType]; // referenced types